src/charm.py

```diff
@@ -0,0 +1,241 @@
+            if isinstance(self.unit.status, ActiveStatus) or self.unit.status.message == "":
+                self.unit.status = WaitingStatus("Waiting for resource limit patch to apply")
+            return
+
+        if not self.container.can_connect():
+            self.unit.status = MaintenanceStatus("Waiting for pod startup to complete")
+            return
+
+        # Make sure the external url is valid
+        if external_url := self._external_url:
+            parsed = urlparse(external_url)
+            if not (parsed.scheme in ["http"] and parsed.hostname):
+                # This shouldn't happen
+                logger.error(
+                    "Invalid external url: %s; must include scheme and hostname.",
+                    external_url,
+                )
+                self.unit.status = BlockedStatus(
+                    f"Invalid external url: '{external_url}'; must include scheme and hostname."
+                )
+                return
+
+        # Update config file
+        try:
+            self.blackbox_workload.update_config()
+        except ConfigUpdateFailure as e:
+            self.unit.status = BlockedStatus(str(e))
+            return
+
+        # Update pebble layer
+        self.blackbox_workload.update_layer()
+
+        # Reload or restart the service
+        self.blackbox_workload.reload()
+
+        self.unit.status = ActiveStatus()
+
+    @property
+    def _internal_url(self) -> str:
+        """Return the fqdn dns-based in-cluster (private) address of the blackbox exporter."""
+        return f"http://{socket.getfqdn()}:{self._port}"
+
+    @property
+    def _external_url(self) -> str:
+        """Return the externally-reachable (public) address of the blackbox exporter.
+
+        If not set in the config, return the internal url.
+        """
+        return self.model.config.get("web-external-url") or self._internal_url
+
+    @property
+    def self_scraping_job(self):
+        """The self-monitoring scrape job."""
+        external_url = urlparse(self._external_url)
+        metrics_path = f"{external_url.path.rstrip('/')}/metrics"
+        target = (
+            f"{external_url.hostname}{':'+str(external_url.port) if external_url.port else ''}"
+        )
+        job = {
+            "metrics_path": metrics_path,
+            "static_configs": [{"targets": [target]}],
+        }
+
+        return [job]
+
+    @property
+    def probes_scraping_jobs(self):
+        """The scraping jobs to execute probes from Prometheus."""
+        jobs = []
+        external_url = urlparse(self._external_url)
+        f"{external_url.path.rstrip('/')}/probe"
+        probes_scrape_jobs = self.model.config.get("probes_file")
+        if probes_scrape_jobs:
+            probes = yaml.safe_load(probes_scrape_jobs)
+            # Add the Blackbox Exporter's `relabel_configs` to each job
+            for probe in probes["scrape_configs"]:
+                probe["relabel_configs"] = [
```

Comment:
(Reminder) Would be great to have a one liner rationale per relabel config.

Reply:
I added some explanation, though I still believe that this is not a good place to explain how `__address__` relabeling works; that's for Prometheus docs to do :D